hail/python/hail/vds/combiner/variant_dataset_combiner.py

```diff
@@ -0,0 +1,445 @@
+
+        paths = [f.path for f in files_to_merge]
+        vdss = read_variant_datasets(paths, intervals, intervals_dtype)
+        combined = combine_variant_datasets(vdss)
+
+        if self.finished:
+            combined.write(self.output_path)
+            return
+
+        new_path = os.path.join(temp_path, 'dataset.vds')
+        combined.write(new_path, overwrite=True)
+        new_n_samples = sum(f.n_samples for f in files_to_merge)
+        self.vdses.append(VDSMetadata(path=new_path, n_samples=new_n_samples))
+        self.vdses.sort(key=lambda x: x.n_samples, reverse=True)
+
+    def _step_gvcfs(self):
+        step = self.branch_factor
+        files_to_merge = self.gvcfs[:self.gvcf_batch_size * step]
+        self.gvcfs = self.gvcfs[self.gvcf_batch_size * step:]
+        if self.gvcf_external_header is not None:
+            sample_names = self.gvcf_sample_names[:self.gvcf_batch_size * step]
+            self.gvcf_sample_names = self.gvcf_sample_names[self.gvcf_batch_size * step:]
+        else:
+            sample_names = None
+        merge_vds = []
+        while files_to_merge:
+            inputs, files_to_merge = files_to_merge[:step], files_to_merge[step:]
+            if sample_names is not None:
+                names, sample_names = sample_names[:step], sample_names[step:]
+            vdses = [transform_gvcf(vcf)
+                     for vcf in hl.import_gvcfs(inputs,
+                                                self.gvcf_import_intervals,
+                                                array_elements_required=False,
+                                                _external_header=self.gvcf_external_header,
+                                                _external_sample_ids=[[name] for name in names] if sample_names is not None else None,
+                                                reference_genome=self.reference_genome,
+                                                contig_recoding=self.contig_recoding)]
+            merge_vds.append(combine_variant_datasets(vdses))
+        if self.finished and len(merge_vds) == 1:
+            merge_vds[0].write(self.output_path)
+            return
+
+        temp_path = self._temp_out_path(f'gvcf-combine_job{self._job_id}/dataset_')
+        pad = len(str(len(merge_vds) - 1))
+        merge_metadata = []
+        count = 0
+        for _ in merge_vds:
+            merge_metadata.append(VDSMetadata(path=temp_path + str(count).rjust(pad, '0') + '.vds',
+                                  n_samples=len(vdses)))
+            count += 1
+        paths = [md.path for md in merge_metadata]
+        hl.vds.write_variant_datasets(merge_vds, paths, overwrite=True)
+        self.vdses.extend(merge_metadata)
+        self.vdses.sort(key=lambda x: x.n_samples, reverse=True)
+
+    def _temp_out_path(self, extra):
+        return os.path.join(self.temp_path, 'combiner-intermidiates', f'{self._uuid}_{extra}')
+
+
+def run_combiner(**kwargs):
```

Comment:
I think we should remove this -- it's hard to document and confusing to have two entry points to creating a new combiner. Can we just add a `run()` method on the combiner it self that does the loop?